### PR TITLE
[red-knot] more ergonomic and efficient handling of known builtin classes

### PR DESCRIPTION
**Pull request by Slyces**
## Summary

This PR introduces a new enumeration `BuiltinType` that serves 2 purposes:
- df16f218aaa40f319fc6dd3eaddefdc7c3b5bfb0: give a common syntax for the convenience shortcuts to get a builtin instance
  - For that purpose, the enum doesn't need to be exhaustive - just to have the types we often need
-  c98ce748f74841eabe73eb1f4dbc77f83454e5f0: save if a class is a builtin on creation, to save time on call when we need to check for custom behaviour (`str(...)`, `bool(...)`, ...)

I think for the first purpose that's mainly a syntax preference, you should be able to tell fairly fast if you prefer it that way.
For the second, that's mainly an optimisation that we might need once we handle more specific behaviour for `builtin` used as callable - this could clearly wait until we implement more of those.

## Test Plan

No tests were added, but the existing test suite is enough to check if this introduced a regression.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types.rs
Comment (dhruvmanila) [inline crates/red_knot_python_semantic/src/types.rs:873]: What's the reason for introducing and using `BuiltinType::iter` here instead of a match. I think I'd prefer to avoid adding a dependency for this use case as it feels a bit overkill :)

Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs]: I guess the reason was to make it easier to add more members as needed - but that's probably not worth a dependency!

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:829]:
```suggestion
/// Feel free to expand this enum if you ever find yourself using the same builtin type in multiple
```

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:862]: maybe

```suggestion
        self.to_class(db).to_instance(db)
```

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:419]: Can we add a method on `ClassType` to make this less verbose? I would probably even call it `is_builtin`:
```suggestion
                if class.is_builtin(db, BuiltinType::Int) =>
```

And then rename the `is_builtin` field to just `builtin`; `is_` prefix suggests a boolean, not an optional enum.

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:1283]: As mentioned above, I'd name this just `builtin` -- `is_` prefix suggests a boolean.

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:832]: Naming nit: let's rename this to `BuiltinClass`. This actually only handles classes, not all kinds of types (functions are types too).

Speaking of which, I kinda want to unify this PR and `FunctionKind`, in terms of both naming and API. The one trick there is that we want to represent known functions that aren't actually builtins (e.g. `reveal_type`). The same may be true in future for other stdlib classes!

So maybe ultimately what we'll actually want is `KnownClass` and `KnownFunction` enums, and a `known` field (which is an `Option<KnownClass>` or `Option<KnownFunction>` - I like this Option approach better than `FunctionKind::Ordinary`), and `is_known` method, on both `ClassType` and `FunctionType`. But this would require generalizing what you have in this PR to not assume builtins module, and instead classify the known functions/classes by module as well. I would be happy with doing all of this now, in this PR, or waiting on it if you prefer to land a simpler version of this PR for now.

cc @AlexWaygood in case you hate my naming/API preferences :)

**Review by carljm — COMMENTED** [review]
This looks great as a general direction! A few nits and thoughts on naming/API.

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:1288]: I don't like the name of this method currently because it implies that if it returns `None` the class is not a builtin. But that's not necessarily true, since our list of known builtins isn't exhaustive. I'd prefer `maybe_known_builtin`. Or just `maybe_known` if we go with the broader refactor/rename to handle any known class, even if it's not a builtin.

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:832]: Hmm, I considered the pros and cons of `Option<KnownFunction>` vs having `FunctionKind::Ordinary` while working on my previous PR. I ended up going with `FunctionKindOrdinary` because I have a general preference having a flat list of possible states rather than representing possible states using an enum of enums. It makes it easier to count exactly how many possible states there are, and giving the default state a name (`Ordinary`) rather than using `None` makes it abundantly clear to new readers of the code what the default state represents. It's also more ergonomic to pattern-match on a flat list rather than a nested enum.

I don't feel _too_ strongly, but a flat enum would be my preference :)

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:832]: Yes, you're definitely right that `FunctionKind` is not a great name. And I think you're right that renaming it to `KnownFunction` addresses some of the concern I had that it wouldn't be obvious what the default state signified. It's pretty obvious that `None` indicates that it's not a known function!

I'm still not a massive fan of nested enums, but I guess in this case it's _okay_. Feel free to proceed :)

Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:832]: Would you like me to address this naming change in this PR?

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:1288]: Completely agree. I think I like this `known` prefix - what would you think of also adding it to the type we store internally?
I'm overall afraid that a new reader would come - see a builtin field - and trust that it covers all existing builtins

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [reply crates/red_knot_python_semantic/src/types.rs:832]: Yes, if you're up for it I think it would be ideal to go ahead and make the switch to both `KnownClass` and `KnownFunction` enums in this PR, with `ClassType` having a field `known: Option<KnownClass>` and a method `is_known`, and similar for `FunctionType`. Where `KnownFunction` would replace the current `FunctionKind` enum, and the new `FunctionType::known` field would replace the current `FunctionType::kind`.

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [reply crates/red_knot_python_semantic/src/types.rs:1288]: Agreed; I think this is a good enough reason to go ahead and fully make the change to "Known" terminology in this PR, as described in the other comment thread.

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs]: The newest commit implements a first attempt at this. I did three things:
- Apply naming convention for `KnownFunction`
- Apply naming convention for `KnownClass`
- Try to extend the structure in `KnownClass` to easily support non-builtins

File: crates/red_knot_python_semantic/src/types.rs
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:419]: Is there a reason you went for `is_known_class` rather than `is_known` for this method? It's ok either way, just feels like an unnecessary level of repetition of the word "class" when we have `class.is_known_class(db, KnownClass::...)`

**Review by carljm — APPROVED** [review]
One question about method naming, otherwise this looks excellent, thank you!

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:419]: No reason, is known is good!

Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs]: @carljm should be fixed!

File: crates/red_knot_python_semantic/src/types/builder.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types/builder.rs:365]: Ah. @carljm, I think this might actually reveal a bug in `main`. The union of `Literal[True] | Literal[False]` should surely simplify to `Instance("bool")` rather than `Class("bool")`. If an object has type `Literal[True]`, it indicates that it the object is an instance of `bool`, not that it is the `bool` class itself.

The fact that the new idiom in this PR makes this bug more obvious is definitely a point in favour of us landing it @Slyces!!

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:758]: Nit: I'd probably add the other "fundamental" types in this list (`types.ModuleType`, `types.FunctionType`, `_typeshed.NoneType` and `builtins.type`) to the `KnownClass` enum as well, and take `typing.Any` out of the enum for now

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:902]: this feels a bit repetitive, and it also feels like we're doing some work that isn't always necessary depending on what `self` is. Maybe something like this?

```suggestion
        let is_stdlib_module = module.search_path().is_standard_library();
        let module_name = module.name().as_str();
        match self {
            Self::Any => is_stdlib_module && matches!(module_name, "typing" | "typing_extensions")
            _ => is_stdlib_module && module_name == "builtins",
        }
```

I'd actually also prefer it if we tediously enumerated all the builtin classes rather than using `_` as a fallback branch. If we do it that way then the compiler will warn us when we add a new non-builtins branch but forget to update this method, which would lead to incorrect results.

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:890]: I wish Rust had some way of enforcing that the r.h.s. of a match statement is meant to be exhaustive :/ I guess that's what you were going for with your `EnumIter` solution from `strum`, but I agree with Dhruv that it's probably not really worth an extra dependency.

If this enum continues to grow and this becomes a maintenance problem, I guess we could consider writing a macro to generate this method, which would allow us to ensure that the r.h.s. of the `match` is exhaustive. But that comes with its own tradeoffs, of course.

**Review by AlexWaygood — COMMENTED** [review]
Seems fine to me other than a couple of nits!

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:902]: Actually we could also early return if we are not in stdlib as all cases in the close future would be in stdlib

File: crates/red_knot_python_semantic/src/types/builder.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types/builder.rs:365]: Great catch, definitely didn't see it myself!

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/builder.rs:365]: (It doesn't need to be fixed in this PR, since I guess it'll require other changes elsewhere that your PR currently doesn't touch! Though it might actually be a pretty small change — if so, it's probably okay to fix it here if you feel like it 😆)

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:758]: Why is it `_typeshed.NoneType`? In my repl I see the `type(None)` from builtins?

Comment (carljm) [reply crates/red_knot_python_semantic/src/types.rs:758]: The type of `None` (`NoneType`) is not exposed in the stdlib at all in some versions of Python, and is never exposed in builtins. The `None` singleton object is in builtins, but typeshed needs a place to give the definition of the _type_.

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:890]: I agree, and indeed this is where my mind went first when I introduced `strum`: make sure that extending the enum is as painless as possible. I'm dropping a `// Note` in there for future reference. Also found a stackoverflow exploring exactly this limitation.

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:758]: Would you have a reference or strategy (maybe to drop in the comment of the enum) to find out where a given common symbol is defined?

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:758]: The [`_typeshed` module](https://github.com/python/typeshed/tree/main/stdlib/_typeshed) is a fictitious module that typeshed pretends is part of the standard library. Because typeshed pretends it's part of the stdlib, and we treat typeshed's stubs as the sole source of truth for the stdlib, we think it's part of the stdlib too!

The purpose of the `_typeshed` module is twofold:
- to define various convenient aliases and protocols that don't exist at runtime but which are used extensively in typeshed's stubs for the stdlib. These need to be defined in a stdlib module, because typeshed's stubs for the stdlib can't depend on any non-stdlib stubs.
- to define certain compatibility workarounds for type checkers (that's us!) to use. `_typeshed.NoneType` is an example of this. At runtime, the type of `None` is `types.NoneType`, but `types.NoneType` was [only added in 3.10](https://docs.python.org/3/library/types.html#types.NoneType). We can't use that as the type of `None`, as we still support Python 3.8 and 3.9, and all other type checkers also have this problem -- so typeshed defines `_typeshed.NoneType` as a little compatibility hack for us.

File: crates/red_knot_python_semantic/src/types/builder.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types/builder.rs:365]: Might prefer a separate PR if that's ok!

Comment (Slyces) [reply crates/red_knot_python_semantic/src/types/builder.rs:365]: https://github.com/astral-sh/ruff/issues/13642

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs:758]: So, small question on the topic. I see that we have methods to lookup `typing_extensions`, but nothing from `typing`

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:758]: > Would you have a reference or strategy (maybe to drop in the comment of the enum) to find out where a given common symbol is defined?

I guess I'd probably `cd` into this directory and grep for the type's name? https://github.com/astral-sh/ruff/tree/main/crates/red_knot_vendored/vendor/typeshed/stdlib

Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:758]: > So, small question on the topic. I see that we have methods to lookup `typing_extensions`, but nothing from `typing`

Yeah, that's for a very similar reason! There's some symbols that only exist in `typing` on some versions of Python, but exist in `typing_extensions` on all versions of Python. That means that it's usually better for a type checker to look things up in `typing_extensions` rather than `typing`, since `typing_extensions` re-exports all symbols from `typing`, and typeshed [pretends `typing_extensions` is part of the stdlib](https://github.com/python/typeshed/blob/main/stdlib/typing_extensions.pyi). (I _promise_ typeshed doesn't do this for _loads_ of modules... only `_typeshed` and `typing_extensions`!!)

**Comment by Slyces** [comment]
If that would be more manageable, one option we would have for the maintainability of mapping `str` to `KnownClass` would be to only add a dependency for tests, which would be used to ensure we didn't miss a case.
Don't know if that's more acceptable, but it would ensure that we never miss a case without adding dependencies to the production code

File: crates/red_knot_python_semantic/src/types.rs
Comment (Slyces) [reply crates/red_knot_python_semantic/src/types.rs]: If that would be more manageable, one option we would have for the maintainability of mapping str to KnownClass would be to only add a dependency for tests, which would be used to ensure we didn't miss a case.
Don't know if that's more acceptable, but it would ensure that we never miss a case without adding dependencies to the production code

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:754]:
```suggestion
            Type::Function(_) => KnownClass::FunctionType.to_class(db),
            Type::Module(_) => KnownClass::ModuleType.to_class(db),
```

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:758]:
```suggestion
            Type::None => KnownClass::NoneType.to_class(db),
            // TODO not accurate if there's a custom metaclass...
            Type::Class(_) => KnownClass::Type.to_class(db),
```

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:843]: nit

```suggestion
    NoneType, // Part of `types` for Python >= 3.10
```

**Review by AlexWaygood — APPROVED** [review]
Thanks!

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs]: Let's land the PR for now. Maybe we can create an issue to discuss this further and see what the other red-knot maintainers think :-)

**Review by MichaReiser — COMMENTED** [review]
I like this a lot. This is excellent API design :)

**Pull request merged**
